defi-pulse/src/adapter.ts

```diff
@@ -0,0 +1,71 @@
+import { Requester, Validator } from '@chainlink/external-adapter'
+import { Execute } from '@chainlink/types'
+import { utils } from 'ethers'
+import { getSymbol } from './symbols'
+import { getAllocations } from './index-allocations'
+import { getPriceAdapter, PriceAdapter } from './priceAdapter'
+import Decimal from 'decimal.js'
+
+export type Index = IndexAsset[]
+
+export type IndexAsset = {
+  asset: string
+  units: Decimal
+  weight: number
+  price?: number
+}
+
+async function makeIndex(components: string[], units: number[], network: string): Promise<Index> {
+  return await Promise.all(
+    components.map(async (component, i) => {
+      return {
+        asset: await getSymbol(component, network),
+        units: new Decimal(new utils.BigNumber(units[i]).toString()).div(1e18),
+        weight: 0,
+      }
+    }),
+  )
+}
+
+const calculateIndexValue = (index: Index): number =>
+  index
+    .reduce((acc, { units, price }) => {
+      if (!price) throw new Error('Invalid price')
+      return acc.plus(units.times(price))
+    }, new Decimal(0))
+    .toNumber()
```

Comment:
Maybe we can split this into an assert and compute parts:

```suggestion
const calculateIndexValue = (index: Index): number => {
  // assert all prices are set
  const isPriceSet = (i: IndexAsset) => (i.price && i.price > 0)
  if (!index.every(isPriceSet)) throw new Error('Invalid index: price not set')
  // calculate total value
  return index
    .reduce((acc, { units, price }) => acc.plus(units.times(price))}, new Decimal(0))
    .toNumber()
}
```

Reply:
I like that more, but he still complains that he might receive an undefined price

Reply:
I've updated the comment. You should be able to force it with a Non-Null Assertion Operator `!`.